omf-python/src/file/reader.rs

```diff
@@ -132,84 +166,235 @@ impl PyReader {
     }
 
     /// Read a Scalar array.
-    pub fn array_scalars(&self, array: &PyScalarArray) -> PyResult<Vec<f64>> {
-        self.0
+    pub fn array_scalars(&self, py: Python<'_>, array: &PyScalarArray) -> PyResult<PyObject> {
+        let scalars = self
+            .0
             .array_scalars(&array.0)
-            .map_err(OmfException::py_err)?
-            .collect::<Result<Vec<_>, _>>()
-            .map_err(OmfException::py_err)
+            .map_err(OmfException::py_err)?;
+        match scalars {
+            omf::data::Scalars::F32(generic_scalars) => Ok(PyArray::from_iter_bound(
+                py,
+                generic_scalars
+                    .collect::<Result<Vec<_>, _>>()
+                    .map_err(OmfException::py_err)?,
+            )
+            .into_py(py)),
+            omf::data::Scalars::F64(generic_scalars) => Ok(PyArray::from_iter_bound(
+                py,
+                generic_scalars
+                    .collect::<Result<Vec<_>, _>>()
+                    .map_err(OmfException::py_err)?,
+            )
+            .into_py(py)),
+        }
     }
 
     /// Read a Vertex array.
-    pub fn array_vertices(&self, array: &PyVertexArray) -> PyResult<Vec<[f64; 3]>> {
-        self.0
+    pub fn array_vertices(&self, py: Python<'_>, array: &PyVertexArray) -> PyResult<PyObject> {
+        let vertices = self
+            .0
             .array_vertices(&array.0)
-            .map_err(OmfException::py_err)?
-            .collect::<Result<Vec<_>, _>>()
-            .map_err(OmfException::py_err)
+            .map_err(OmfException::py_err)?;
+        match vertices {
+            omf::data::Vertices::F32(generic_arrays) => {
+                let vec: Vec<_> = generic_arrays
+                    .collect::<Result<_, _>>()
+                    .map_err(OmfException::py_err)?;
+
+                let shape = (vec.len(), 3);
+
+                // TODO: remove unwrap
+
+                Ok(PyArray::from_owned_array_bound(
+                    py,
+                    Array::from_shape_vec(shape, vec.into_flattened()).unwrap(),
+                )
+                .into_py(py))
```

Comment:
Don't repeat yourself. Module level-functions for converting flat, wide (like this one), nullable-flat, and nullable-wide arrays should tidy this up.

Reply:
Yep, I’m planning to move all the repeated code into methods.